sdk/python/kfp/dsl/_container_op.py

```diff
@@ -12,309 +12,940 @@
-
-    return self.add_resource_limit("%s.com/gpu" % vendor, gpu)
-
-  def add_volume(self, volume):
-    """Add K8s volume to the container
-
-    Args:
-      volume: Kubernetes volumes
-      For detailed spec, check volume definition
-      https://github.com/kubernetes-client/python/blob/master/kubernetes/client/models/v1_volume.py
+        self.volume_mounts = create_and_append(self.volume_mounts,
+                                               volume_mount)
+        return self
+
+    def add_volume_devices(self, volume_device):
+        """
+        Add a block device to be used by the container.
+
+        Args:
+          volume_device: Kubernetes volume device
+          For detailed spec, volume device definition
+          https://github.com/kubernetes-client/python/blob/master/kubernetes/client/models/v1_volume_device.py
+        """
+
+        if not isinstance(volume_device, V1VolumeDevice):
+            raise ValueError(
+                'invalid argument. Must be of instance `V1VolumeDevice`.')
+
+        self.volume_devices = create_and_append(self.volume_devices,
+                                                volume_device)
+        return self
+
+    def add_env_variable(self, env_variable):
+        """Add environment variable to the container.
+
+        Args:
+          env_variable: Kubernetes environment variable
+          For detailed spec, check environment variable definition
+          https://github.com/kubernetes-client/python/blob/master/kubernetes/client/models/v1_env_var.py
+        """
+
+        if not isinstance(env_variable, V1EnvVar):
+            raise ValueError(
+                'invalid argument. Must be of instance `V1EnvVar`.')
+
+        self.env = create_and_append(self.env, env_variable)
+        return self
+
+    def add_env_from(self, env_from):
+        """Add a source to populate environment variables int the container.
+
+        Args:
+          env_from: Kubernetes environment from source
+          For detailed spec, check environment from source definition
+          https://github.com/kubernetes-client/python/blob/master/kubernetes/client/models/v1_env_var_source.py
+        """
+
+        if not isinstance(env_from, V1EnvFromSource):
+            raise ValueError(
+                'invalid argument. Must be of instance `V1EnvFromSource`.')
+
+        self.env_from = create_and_append(self.env_from, env_from)
+        return self
+
+    def set_image_pull_policy(self, image_pull_policy):
+        """Set image pull policy for the container.
+
+        Args:
+          image_pull_policy: One of `Always`, `Never`, `IfNotPresent`. 
+        """
+        if image_pull_policy not in ['Always', 'Never', 'IfNotPresent']:
+            raise ValueError(
+                'Invalid imagePullPolicy. Must be one of `Always`, `Never`, `IfNotPresent`.'
+            )
+
+        self.image_pull_policy = image_pull_policy
+        return self
+
+    def add_port(self, container_port):
```

Comment:
Maybe add a TODO to support pipelineparam for the other configurations besides CPU, memory ,etc.

Reply:
It is already supported - in the `compiler._op_to_template` there is a recursive method to replace `PipelineParam` and its serialized form into argo variables.